File: kt-lox/src/main/kotlin/com/craftinginterpreters/lox/Lox.kt
package com.craftinginterpreters.lox

import java.io.File
import kotlin.system.exitProcess

object Lox {
    private var hadError: Boolean = false
    private var hadRuntimeError: Boolean = false
    private val interpreter = Interpreter()

    fun runFile(path: String) {
        val contents = File(path).readText()
        run(contents)

        if (hadError) exitProcess(65)
        if (hadRuntimeError) exitProcess(70)
    }

    fun runPrompt() {
        while (true) {
            print("> ")

            val line = readlnOrNull() ?: return
            if (line == "exit") return
            run(line)

            hadError = false
            hadRuntimeError = false
        }
    }

    private fun run(source: String) {
        val scanner = Scanner(source)
        val tokens = scanner.scanTokens()

        val parser = Parser(tokens)
        val statements = parser.parse()
        if (hadError) return

        val resolver = Resolver(interpreter)
        resolver.resolve(statements)
        if (hadError) return

        interpreter.interpret(statements)
    }

    fun error(line: Int, message: String) {
        report(line, "", message)
    }

    fun error(token: Token, message: String) {
        if (token.type == TokenType.EOF) {
            report(token.line, " at end", message)
        } else {
            report(token.line, " at '${token.lexeme}'", message)
        }
    }

    private fun report(line: Int, where: String, message: String) {
        System.err.println("[line $line] Error $where: $message")
        hadError = true
    }

    fun runtimeError(error: RuntimeError) {
        System.err.println("[line ${error.token.line}] Error: ${error.message}")
        hadRuntimeError = true
    }
}

File: rs-lox/src/lox.rs
use anyhow::{bail, Result};

use crate::scanner::tokenize;

#[derive(Debug)]
pub struct Lox {
    pub had_error: bool,
}

impl Lox {
    pub fn new() -> Self {
        Self { had_error: false }
    }
    pub fn run(&mut self, source: &str) -> Result<()> {
        let (tokens, errors) = tokenize(source);
        dbg!(tokens);
        self.had_error = !errors.is_empty();
        errors
            .into_iter()
            .for_each(|(line, message)| self.error(line, &message));
        if self.had_error {
            bail!("Lox encountered errors");
        }
        Ok(())
    }

    pub(crate) fn error(&self, line: usize, message: &str) {
        self.report(line, "", message);
    }

    pub(crate) fn report(&self, line: usize, loc: &str, message: &str) {
        println!("[line {line}] Error {loc}: {message}")
    }
}
